### PR TITLE
JAMES-2411 Allow to download an EML in the webadmin repository endpoint

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java [hunk 1]
```diff
@@ -177,25 +181,50 @@ public void defineGetMailRepositories() {
...
-                    .message("Error while retrieving mail")
-                    .haltError();
-            }
+
+        service.get(MAIL_REPOSITORIES + "/:encodedUrl/mails/:mailKey", Constants.JSON_CONTENT_TYPE, (request, response) -> {
+            return getMailAsJson(decodedRepositoryUrl(request), request.params("mailKey"));
         }, jsonTransformer);
+
+        service.get(MAIL_REPOSITORIES + "/:encodedUrl/mails/:mailKey", Constants.RFC822_CONTENT_TYPE, (request, response) -> {
+            response.type(Constants.RFC822_CONTENT_TYPE);
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: Do you miss a `Content-Length` header?

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/MailRepositoryStoreService.java [hunk 2]
```diff
@@ -83,6 +86,14 @@ public MailRepositoryStoreService(MailRepositoryStore mailRepositoryStore) {
             .map(Throwing.function(MailDto::fromMail).sneakyThrow());
     }
 
+    public Optional<InputStream> downloadMail(String url, String mailKey) throws MailRepositoryStore.MailRepositoryStoreException, MessagingException {
+        MailRepository mailRepository = getRepository(url);
+
+        return Optional.ofNullable(mailRepository.retrieve(mailKey))
+            .map(Throwing.function(Mail::getMessage).sneakyThrow())
+            .map(Throwing.function(MimeMessage::getRawInputStream).sneakyThrow());
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/MailRepositoryStoreService.java]: From my experience getRawInputStream returns just partial data

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 3]
```diff
@@ -465,6 +469,54 @@ public void retrievingAMailShouldFailWhenUnknown() throws Exception {
...
+        when(mailRepositoryStore.get(URL_MY_REPO)).thenReturn(Optional.of(mailRepository));
+
+        String name = NAME_1;
+        FakeMail mail = FakeMail.builder()
+            .name(name)
+            .fileName("mail.eml")
+            .build();
+        mailRepository.store(mail);
+
+        String expectedContent = IOUtils.toString(mail.getMessage().getRawInputStream(), StandardCharsets.UTF_8);
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: Sad... You don't really test things and fail to recognise you do return partial data.

Here you should read data from the file.

**Comment by chibenwa** [comment]
```
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Results :
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] 
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Failed tests: 
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27]   MailRepositoriesRoutesTest.downloadingAMailShouldReturnTheEml:494 1 expectation failed.
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Expected header "Content-Length" was not "552", was "547". Headers are:
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Date=Thu, 07 Jun 2018 03:32:32 GMT
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Access-Control-Allow-Origin=*
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Access-Control-Request-Method=DELETE, GET, POST, PUT
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Access-Control-Allow-Headers=Content-Type, Authorization, Accept
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Content-Type=message/rfc822
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Content-Length=547
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Server=Jetty(9.4.7.v20170914)
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] 
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27]   MailRepositoryStoreServiceTest.dowloadMailShouldReturnThInputStreamWhenMailExists:176 expected:<...: <general@list.org>[
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] MIME-Version: 1.0
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Received: by 10.27.96.135 with HTTP; Wed, 3 Jun 2015 21:08:41 -0700 (PDT)
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] From: Hello <hello@gmail.com>
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Date: Thu, 4 Jun 2015 06:08:41 +0200
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Message-ID: <CAN1zdnW5PW7z5msM-TjUpouygxWY6z-25A3WCB-C5BvqDuP1Qw@mail.gmail.com>
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] To: General Discussion  <general@list.org>
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Content-Type: text/plain; charset=UTF-8
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Subject: [arch-general] Very important topic
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Reply-To: General Discussion <general@list.org>
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Sender: "arch-general" <general@list.org>
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] ]
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Hi all,
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] 
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Very impor...> but was:<...: <general@list.org>[
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] MIME-Version: 1.0
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Received: by 10.27.96.135 with HTTP; Wed, 3 Jun 2015 21:08:41 -0700 (PDT)
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] From: Hello <hello@gmail.com>
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Date: Thu, 4 Jun 2015 06:08:41 +0200
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Message-ID: <CAN1zdnW5PW7z5msM-TjUpouygxWY6z-25A3WCB-C5BvqDuP1Qw@mail.gmail.com>
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] To: General Discussion  <general@list.org>
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Content-Type: text/plain; charset=UTF-8
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Subject: [arch-general] Very important topic
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Reply-To: General Discussion <general@list.org>
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Sender: "arch-general" <general@list.org>
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] 
]
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Hi all,
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] 
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Very impor...>
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] 
[e7c50f803fcd87c34b7acc33b59ea4a2614c4b27] Tests run: 66, Failures: 2, Errors: 0, Skipped: 0
```

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/service/MailRepositoryStoreServiceTest.java [hunk 4]
```diff
@@ -126,4 +132,47 @@ public void listMailsShouldApplyLimitAndOffset() throws Exception {
...
+    @Test
+    public void dowloadMailShouldReturnEmptyWhenMailNotFound() throws Exception {
+        when(mailRepositoryStore.get(FIRST_REPOSITORY)).thenReturn(Optional.of(repository));
+
+        assertThat(testee.retrieveMessage(FIRST_REPOSITORY, NAME_1))
+            .isEmpty();
+    }
+
+    @Test
+    public void dowloadMailShouldReturnThInputStreamWhenMailExists() throws Exception {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/service/MailRepositoryStoreServiceTest.java]: Th ?

**Comment by chibenwa** [comment]
test this please (because anakin is slooooow)

**Comment by aduprat** [comment]
:+1:

**Comment by aduprat** [comment]
:+1:

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java [hunk 5]
```diff
@@ -169,33 +176,72 @@ public void defineGetMailRepositories() {
...
+    private Object writeMimeMessage(MimeMessage mimeMessage, HttpServletResponse rawResponse) throws MessagingException, IOException {
+        rawResponse.setContentType(Constants.RFC822_CONTENT_TYPE);
+        rawResponse.setHeader("Content-Length", String.valueOf(computeExactSize(mimeMessage)));
+        mimeMessage.writeTo(rawResponse.getOutputStream());
+        return rawResponse;
+    }
+
+    private long computeExactSize(MimeMessage mimeMessage) throws IOException, MessagingException {
+        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+        mimeMessage.writeTo(byteArrayOutputStream);
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: do we really need to load the message in memory to compute its size?

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: Yes: we should know the size prior to the download to start and the getSize method is not accurate (I had test failures)

Comment (rouazana) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: ok that's not what I meant, but let's stick with that

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**